% Template for PLoS
% Version 3.1 February 2015
%
% To compile to pdf, run:
% latex plos.template
% bibtex plos.template
% latex plos.template
% latex plos.template
% dvipdf plos.template
%
% % % % % % % % % % % % % % % % % % % % % %
%
% -- IMPORTANT NOTE
%
% This template contains comments intended 
% to minimize problems and delays during our production 
% process. Please follow the template instructions
% whenever possible.
%
% % % % % % % % % % % % % % % % % % % % % % % 
%
% Once your paper is accepted for publication, 
% PLEASE REMOVE ALL TRACKED CHANGES in this file and leave only
% the final text of your manuscript.
%
% There are no restrictions on package use within the LaTeX files except that 
% no packages listed in the template may be deleted.
%
% Please do not include colors or graphics in the text.
%
% Please do not create a heading level below \subsection. For 3rd level headings, use \paragraph{}.
%
% % % % % % % % % % % % % % % % % % % % % % %
%
% -- FIGURES AND TABLES
%
% Please include tables/figure captions directly after the paragraph where they are first cited in the text.
%
% DO NOT INCLUDE GRAPHICS IN YOUR MANUSCRIPT
% - Figures should be uploaded separately from your manuscript file. 
% - Figures generated using LaTeX should be extracted and removed from the PDF before submission. 
% - Figures containing multiple panels/subfigures must be combined into one image file before submission.
% For figure citations, please use "Fig." instead of "Figure".
% See http://www.plosone.org/static/figureGuidelines for PLOS figure guidelines.
%
% Tables should be cell-based and may not contain:
% - tabs/spacing/line breaks within cells to alter layout or alignment
% - vertically-merged cells (no tabular environments within tabular environments, do not use \multirow)
% - colors, shading, or graphic objects
% See http://www.plosone.org/static/figureGuidelines#tables for table guidelines.
%
% For tables that exceed the width of the text column, use the adjustwidth environment as illustrated in the example table in text below.
%
% % % % % % % % % % % % % % % % % % % % % % % %
%
% -- EQUATIONS, MATH SYMBOLS, SUBSCRIPTS, AND SUPERSCRIPTS
%
% IMPORTANT
% Below are a few tips to help format your equations and other special characters according to our specifications. For more tips to help reduce tshe possibility of formatting errors during conversion, please see our LaTeX guidelines at http://www.plosone.org/static/latexGuidelines
%
% Please be sure to include all portions of an equation in the math environment.
%
% Do not include text that is not math in the math environment. For example, CO2 will be CO\textsubscript{2}.
%
% Please add line breaks to long display equations when possible in order to fit size of the column. 
%
% For inline equations, please do not include punctuation (commas, etc) within the math environment unless this is part of the equation.
%
% % % % % % % % % % % % % % % % % % % % % % % % 
%
% Please contact latex@plos.org with any questions.
%
% % % % % % % % % % % % % % % % % % % % % % % %

\documentclass[10pt,letterpaper]{article}
\usepackage[top=0.85in,left=2.75in,footskip=0.75in]{geometry}

% Use adjustwidth environment to exceed column width (see example table in text)
\usepackage{changepage}

% Use Unicode characters when possible
\usepackage[utf8]{inputenc}

% textcomp package and marvosym package for additional characters
\usepackage{textcomp,marvosym}

% fixltx2e package for \textsubscript
\usepackage{fixltx2e}

% amsmath and amssymb packages, useful for mathematical formulas and symbols
\usepackage{amsmath,amssymb}

% cite package, to clean up citations in the main text. Do not remove.
\usepackage{cite}

% Use nameref to cite supporting information files (see Supporting Information section for more info)
\usepackage{nameref,hyperref}

% line numbers
\usepackage[right]{lineno}

% ligatures disabled
\usepackage{microtype}
\DisableLigatures[f]{encoding = *, family = * }

% rotating package for sideways tables
\usepackage{rotating}

% Remove comment for double spacing
%\usepackage{setspace} 
%\doublespacing

% Text layout
\raggedright
\setlength{\parindent}{0.5cm}
\textwidth 5.25in 
\textheight 8.75in

% Bold the 'Figure #' in the caption and separate it from the title/caption with a period
% Captions will be left justified
\usepackage[aboveskip=1pt,labelfont=bf,labelsep=period,justification=raggedright,singlelinecheck=off]{caption}

% Use the PLoS provided BiBTeX style
\bibliographystyle{plos2015}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother

% Leave date blank
\date{}

% Header and Footer with logo
\usepackage{lastpage,fancyhdr,graphicx}
\usepackage{epstopdf}
\pagestyle{myheadings}
\pagestyle{fancy}
\fancyhf{}
\lhead{\includegraphics[width=2.0in]{PLOS-submission.eps}}
\rfoot{\thepage/\pageref{LastPage}}
\renewcommand{\footrule}{\hrule height 2pt \vspace{2mm}}
\fancyheadoffset[L]{2.25in}
\fancyfootoffset[L]{2.25in}
\lfoot{\sf PLOS}

%% Include all macros below

\newcommand{\lorem}{{\bf LOREM}}
\newcommand{\ipsum}{{\bf IPSUM}}
\newcommand{\PROTEINLIMIT}{600}
\newcommand{\ARCHBACTERTOTAL}{683}
\DeclareMathOperator*{\argmin}{arg\,min}
\DeclareMathOperator*{\argmax}{arg\,max}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%   THIS IS FOR HIGHLIGHTING - TO BE DELETED PRIOR SUBMISSION  %%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\usepackage{color,soul}
\sethlcolor{yellow}




\begin{document}
\vspace*{0.35in}


% Title must be 150 characters or less
\begin{flushleft}
{\Large
\textbf{Thermophilic adaptation in prokaryotes is constrained by metabolic costs of proteostasis}
}
\newline
% Insert Author names, affiliations and corresponding author email.
\\
Sergey V. Venev, 
Konstantin B. Zeldovich\textsuperscript{*}
\\
\bigskip
\bf{} Program in Bioinformatics and Integrative Biology, University of Massachusetts Medical School, Worcester, MA, USA
\\
\bigskip
* konstantin.zeldovich@umassmed.edu
\end{flushleft}

% Please keep the abstract between 250 and 300 words
\section*{Abstract}
Prokaryotes evolved to thrive in an extremely diverse set of habitats, and their proteomes bear signatures of environmental conditions. Although correlations between amino acid usage and environmental temperature are well documented, the underlying molecular mechanisms of thermal adaptation remain poorly understood.  Here, we couple the energetic costs of protein folding and protein homeostasis to build a microscopic model explaining both the overall amino acid composition and its temperature trends. Low biosynthesis costs lead to low diversity of physical interactions between amino acid residues, which in turn makes proteins less stable and drives up chaperone activity to maintain appropriate levels of folded, functional proteins.  Minimization of total energy expenditure defines optimum amino acid frequencies for a given temperature and cost of chaperone activity. Assuming that the cost of chaperone activity is proportional to the fraction of unfolded client proteins at a given temperature, we simulated thermal adaptation of lattice proteins subject to minimization of protein homeostasis costs. For the first time, we were able to predict both the proteome-average amino acid abundances and their temperature trends simultaneously, and found strong correlations between model predictions and 683 genomes of bacteria and archaea. Furthermore, the energetic constraint on protein evolution is more apparent in highly expressed proteins, selected by codon adaptation index. We found that in bacteria, highly expressed proteins are similar in composition to thermophilic ones, while in archaea no correlation between predicted expression level and thermostability was observed.
At the same time, thermal adaptations of highly expressed proteins in bacteria and archaea are nearly identical, suggesting that universal energetic constraints prevail over the phylogenetic differences between these domains of life.

% Please keep the Author Summary between 150 and 200 words
% Use first person. PLOS ONE authors please skip this step. 
% Author Summary not valid for PLOS ONE submissions.   
\section*{Author Summary}

Bacteria and archaea evolved to thrive in habitats ranging from permafrost to hot springs and ocean vents. The remarkable adaptations to these wildly varying environments are imprinted in their genomes. In particular, thermophilic organisms have a specific pattern of amino acid usage, distinguishing them from other species. Insights into why amino acid frequencies in prokaryotic genomes evolved to their present levels, and why do thermophilic organisms have a preference for specific amino acid types are crucial for establishing the basic biophysical constraints of evolution. We propose that amino acid frequencies are set by the limited amount of energy a cell can spend on protein synthesis and repair of defective, misfolded proteins by chaperone molecules. Extensive use of “cheap” amino acids leads to unstable proteins, driving up repair costs. 
Therefore, there exists an amino acid composition that minimizes the total energy spent on protein synthesis and repair. Using simulations, we predicted amino acid frequencies favored at different environmental temperatures, and found a strong agreement between the model and bioinformatics data on 683 genomes. For proteins produced by the cells in largest quantities, today’s bacteria and archaea show identical adaptation patterns, demonstrating a universal energetic constraint that persists since the two domains of life have split billions years ago.

\section*{Introduction}

Over the 4 billion years of evolution, life has colonized an extreme diversity of physical environments on Earth, ranging from volcanic hot vents in the oceans to permafrost to hypersaline lakes. Adaptations to these conditions allow proteins and nucleic acids to function in a wide range of physical and chemical environments, resulting in specific patterns of nucleotide and amino acid usage~\cite{Galtier1997Relationships,Kreil2001Identification,Zeldovich2007Protein,Berezovsky2007Positive,England2003Natural,Fukuchi2003Unique,Sghaier2013There,Sabath2013Growth}. Although the variation of amino acid frequencies across species is relatively constrained for a given genomic  composition~\cite{Krick2014Amino,Goncearenco2014Fundamental}, amino acid compositions of prokaryotic proteomes are sensitive to the temperature and salinity of their natural environments~\cite{Fukuchi2003Unique,Kreil2001Identification}. Unraveling the evolutionary origins of  amino acids usage in proteins, as well as their adaptations to environmental conditions would greatly enhance our understanding of the link between molecular and organismal evolutionary scales, and relative importance of biological and physical factors as selective pressures.

Rationalizing amino acid usage in extant life involves two main questions, first, what are the origins of the generally similar average amino acid usage across multiple highly divergent species, and second, what biological mechanisms drive adaptation of amino acid frequencies to environmental conditions. 

Correlations between nucleotide and amino acid frequencies have been revealed simultaneously with the discovery of the genetic code~\cite{Sueoka1961Correlation}, and the genetic code largely explains observed patterns~\cite{Jukes1975Amino,King1969NonDarwinian}. Subsequent genome-wide studies found that the genomic composition strongly affects the patterns of amino acid and codon usage at the organismal level~\cite{Kreil2001Identification,Knight2001Simple,Lightfield2011Across,Goncearenco2014Fundamental}. At the same time, mutational patterns cannot fully explain the genome composition~\cite{Rocha2010Mutational}. Closely related species adapted to different environments demonstrate variation in amino acid usage unaccounted for by their similar genomic compositions~\cite{Singer2003Thermophilic,Mcdonald2010Temperature,Haney1999Thermal,Fukuchi2003Unique}. 

Advances in understanding protein folding and function, as well as the energetic and biochemical details of cellular metabolism have led to multiple hypotheses about how protein-level selection may affect amino acid composition. In unicellular organisms, highly abundant proteins have a biased amino acid composition to decrease the metabolic cost of amino acid biosynthesis~\cite{Akashi2002Metabolic,Heizer2006Amino}.  Proteome-wide, amino acid usage is negatively correlated with amino acids biosynthesis costs~\cite{Seligmann2003CostMinimization,Heizer2011Amino}. In this class of models, the best explanation of observed amino acid compositions is achieved in a phenomenological approach by Krick et al~\cite{Krick2014Amino}, who took into account the metabolic cost of amino acids synthesis and the rates of their chemical degradation.  As those models do not offer a straightforward way of introducing environmental temperature, they are difficult to extend for modeling thermal adaptation, an established factor affecting amino acid compositions~\cite{Zeldovich2007Protein,Singer2003Thermophilic,Kreil2001Identification,Haney1999Thermal}.

Thermal adaptation in prokaryotes is manifested on many levels, including lipids composition of cell membranes~\cite{Chugunov2014Liquid}, improved stability of functional RNAs~\cite{Galtier1997Relationships}, purine loading of the genes~\cite{Lambros2003Optimum},  genome--level optimizations~\cite{Sabath2013Growth,Saha2015Overlapping},  and specific protein structures~\cite{Szilagyi2000Structural,England2003Natural}. The temperature span of life reaches almost 120$K$ (from -10\textcelsius\ to 110\textcelsius), a change in energy of 0.24 kcal/mol. As this value is comparable to the average effect of a single amino acid substitution in a folded protein $\Delta\Delta G\approx 1$ kcal/mol~\cite{Zeldovich2007Proteinb} and the typical energy of inter--residue van der Waals contacts, thermophilic proteins evolved sequence and structure features to increase their stability. Thermally adapted proteins utilize positive and negative design strategies, stabilizing their native folds and destabilizing unfolded conformations~\cite{Berezovsky2007Positive}. Increased fraction of hydrophobic residues contributes to protein core stability, while increased fraction of the charged residues enforces native fold uniqueness by destabilizing unfolded conformations. \hl{The latter can be achieved by increasing fraction of charged residues on protein surface and formation of ionic pairs}~\cite{Szilagyi2000Structural,Zhao2011Charged}).


Currently, theoretical models are able to explain either the overall proteomic amino acid composition~\cite{Seligmann2003CostMinimization,Heizer2011Amino,Krick2014Amino}, or its temperature trends~\cite{Berezovsky2007Positive,Venev2015Massively}, but not both. Here, we couple protein folding and protein homeostasis costs to bridge this gap and build a microscopic model explaining both amino acid composition and its temperature trends. As it is known, chaperone--assisted folding mechanisms evolved to repair misfolded proteins~\cite{Hartl2011Molecular}, and even a moderate decrease in protein foldability imposes an organismal fitness cost~\cite{Drummond2008MistranslationInduced,Samerotte2011Misfolded}. Chaperones require energy to function, which in turn creates an additional selective pressure on protein foldability, especially in the case of highly abundant proteins~\cite{Kepp2014Model}. As proteostasis consumes up to 80\% of total metabolic rate of unicellular free--living organisms~\cite{Kepp2014Model}, adaptation towards energy efficiency is a significant driver of sequence evolution. We propose that amino acid composition evolved under the selective pressure of the total energetic cost of proteostasis. Following earlier studies, our model includes the cost of \hl{amino acid synthesis and maintainance of their constant concentrations} in the presence of chemical degradation, Fig.~\ref{fig:fig1}. The key new feature, however, comes from considering the energy cost of chaperone assisted protein folding. Protein stability against thermal unfolding depends on the amino acid composition, and amino acid compositions delivering highly foldable proteins require lower energy expenditures on repairing misfolded proteins by chaperones. As detailed below, minimization of the total energy spent on amino acid synthesis and maintenance of folded proteins by chaperones provides an accurate description of both average amino acid frequencies, and their trends with environmental temperature.

% You may title this section "Methods" or "Models". 
% "Models" is not a valid title for PLoS ONE authors. However, PLoS ONE
% authors may use "Analysis" 
% old name "Materials and Methods"
\section*{Models}

Our model for protein homeostasis costs closely follows Kepp~\cite{Kepp2014Model}, together with the hypothesis that amino acid composition generates an additional, folding-related energy demand due to chaperone activity required to refold misfolded or unstable proteins. The cellular energy balance equation can be written as
\begin{equation}
	\label{cell_energy_balance}
	\mathcal{E}_{t} = \mathcal{E}_{m} + \mathcal{E}_{r}
\end{equation}
where $\mathcal{E}_{t}$ is the total energy extracted by the cell from food and recycling per unit time (metabolic rate), this energy is being spent on maintenance of cellular machinery $\mathcal{E}_{m}$ and reproduction $\mathcal{E}_{r}$. As in~\cite{Kepp2014Model}, we assume that cellular fitness is proportional to the amount of energy spent on replication $\mathcal{E}_{r}$.  Since a large fraction of energy is spent on homeostasis, especially in the case of prokaryotes~\cite{Harold1987Vital}, we neglect all other maintenance costs except protein homeostasis cost, so $\mathcal{E}_{m}\approx\mathcal{E}_{p}$.
The energy balance equation~\eqref{cell_energy_balance} confers fitness advantage to the cells with optimized proteostasis, as it allows them to spend more energy on reproduction given equal metabolic rate $\mathcal{E}_{t}$. 

In order to quantify proteostasis costs, we consider three major pathways: protein synthesis, chaperone assisted protein folding, and protein degradation/recycling,  Fig.~\ref{fig:fig1}. Thus, proteostasis energy cost can be written as
\begin{equation}
	\label{proteostasis_cost_expansion}
	\mathcal{E}_{p} \approx \mathcal{E}_{s} + \mathcal{E}_{f} + \mathcal{E}_{d},
\end{equation}
where subscripts $s,f,d$ denote synthesis, folding and degradation respectively. We assume that protein synthesis cost $\mathcal{E}_{s}$ depends on the protein sequence only via amino acid composition and sequence length $L$, i.e., for simplicity we assume equal translation efficiency for all sequences, disregarding codon usage details. We also assume that the pool of amino acids is maintained \hl{in a} steady state condition (constant concentration of each type), so amino acids molecules consumed during protein synthesis are being replaced with newly synthesized or recycled ones. Thus, the total protein synthesis cost $\mathcal{E}_{s}$ can be approximated by
\begin{equation}
	\label{synthesis_cost}
	\mathcal{E}_{s} \approx \sigma L + \sum\limits_{a=1}^{20}C_{a}n_{a},
\end{equation}
where the first term is the cost of translating $L$ codons ($\sigma$ per each), and the second term represents the total energy of synthesizing $n_{a}$ amino acids of each kind in a protein, $\sum\limits_{i=1}^{20}n_{a} = L$. The vector $C_{a}, a=1,\dots,20$ in equation~\eqref{synthesis_cost} is the amount of energy required per unit time to maintain a constant concentration of each type of amino acid monomers, as they are consumed by protein synthesis and also being chemically degraded at different rates, as derived in~\cite{Krick2014Amino}. 
 

\begin{figure}[h!]
\includegraphics[width=\textwidth]{Fig1.png}
\caption{
{\bf Material and energy flux in proteostasis.}  Amino acid biosynthesis, translation and polypeptide synthesis, and chaperone assisted protein folding consume a significant fraction of energy $\mathcal{E}_{t}$ available to a prokaryote. Maintenance of steady state concentrations of every amino acid bears a known energy cost, with cheaper amino acids preferred in highly expressed proteins~\cite{Akashi2002Metabolic}. We propose that energy cost of chaperone activity depends on amino acid composition of client proteins, as protein foldability is affected by amino acid composition~\cite{Dill1985Theory,Berezovsky2007Positive,Venev2015Massively}. Therefore, amino acid composition evolves under the energetic constraint from two distinct processes, amino acid biosynthesis costs and chaperone activity.
}
\label{fig:fig1}
\end{figure}


Maintenance of constant concentration of folded, functional proteins involves the action of chaperones, which help refold improperly folded proteins. Chaperone-assisted refolding consumes energy primarily on conformational transitions required to form the hydrophobic cavity~\cite{Hartl2011Molecular}. We assume that the cost of chaperone activity is proportional to the fraction of unfolded client proteins, which in turn depends on the amino acid composition of the client. Assuming \hl{the fraction of chaperone clients in the native state is $P_{nat}$ ...} that the probability of a chaperone client protein to be natively folded is $P_{nat}$, the energy costs of chaperone activity and protein degradation can be expressed as 
\begin{equation}
	\label{chaperone_degradation_cost}
	\mathcal{E}_{f} + \mathcal{E}_{d} \approx (F+D_{F})\cdot P_{nat} + (U+D_{U})\cdot\left(1-P_{nat}\right)
\end{equation}
where $F$ is the energy spent per unit time to assist successful folding of the $P_{nat}$ fraction of natively-folded protein, $U$ is energy consumed by chaperones to refold $1-P_{nat}$ fraction of client proteins that fail to fold spontaneously, and $D_F$, $D_U$ are proteasome energy expenditures of degrading natively--folded and non--natively--folded proteins, respectively. We assume that $F+D_{F}<U+D_{U}$, i.e. maintenance of misfolded proteins is costlier than maintenance of the well--folding ones. This assumption is supported by the evidence of dosage--dependent fitness penalty induced by misfolding mutations in a protein unrelated to cellular metabolism~\cite{Samerotte2011Misfolded}. As we describe proteostasis as a steady state phenomenon, temporal effects such as differences in protein and amino acid molecule lifetime, as well as kinetics of protein folding and refolding are not considered.


To model folding of chaperone client proteins, we \hl{used} a lattice model~\cite{Shakhnovich1990Enumeration,Sikosek2014Biophysics} of compact polymers on a $4\times4\times4$ cubic lattice, with a randomly generated subset of $N=10^{4}$ conformations. We \hl{used} energy of non--local contacts of a given sequence $S$ threaded onto all $N$ conformations to determine the native state of $S$ as the conformation with the lowest contact energy $E_{nat}$. A residue level knowledge--based potential~\cite{Miyazawa1999SelfConsistent} $\epsilon_{a,b},\, a,b=1,\dots,20$ \hl{was} used to calculate energy of non--local contact in each conformation:
\begin{equation}
	\label{protein_globule_energy}
	E_{i} = \sum\limits_{k,l=1}^{L}\epsilon_{S_{k}S_{l}}\delta^{i}_{kl}
\end{equation}
where $i=1,\dots,N$ is the index of conformation, $S_{k}$ is the type of the amino acid at position $k$ of the sequence $S$, and $\delta^{i}_{kl}$ is a contact map of the conformation $i$, $\delta^{i}_{kl}=1$ if residues $k$ and $l$ are in contact ($|k-l|>1$), and $\delta^{i}_{kl}=0$ otherwise. The equilibrium fraction of natively--folded proteins $P_{nat}$ \hl{was} calculated from the Boltzmann distribution:
\begin{equation}
	\label{pnat_boltzmann}
	P_{nat} = \frac{\exp\left(-E_{nat}/k_{B}T\right)}{\sum\limits_{i=1}^{N}\exp\left(-E_{i}/k_{B}T\right)}
\end{equation}
where $k_{B}$ is the Boltzmann constant. 


Substituting equations \eqref{synthesis_cost},\eqref{chaperone_degradation_cost} \hl{into} equation \eqref{proteostasis_cost_expansion}, we \hl{derived} the proteostasis cost, 
\begin{equation}
	\label{proteostasis_cost_detailed}
	\mathcal{E}_{p} \approx \sigma L + \sum\limits_{a=1}^{20}C_{a}n_{a} + (F+D_{F})\cdot P_{nat} + (U+D_{U})\cdot\left(1-P_{nat}\right)
\end{equation}
which can be further simplified to:
\begin{equation}
	\label{proteostasis_cost_simplified}
	\mathcal{E}_{p} = \alpha - \beta\left(P_{nat} - w\cdot\sum\limits_{a=1}^{20}C_{a}n_{a} \right)
\end{equation}
where $\alpha$,\, $\beta > 0$ and $w>0$, are constants within our framework, and $C_{a}$ is the vector of amino acid maintenance costs. Importantly, foldability $P_{nat}$ of chaperone clients depends on their amino acid composition, allowing for a nontrivial interplay between the two terms in~\eqref{proteostasis_cost_simplified}. The parameter $w$ controls the relative fitness costs of protein (mis)folding (implicit via chaperone activity) and amino acid biosynthesis in our model. The limiting case of $w=0$ corresponds to a purely physical model where fitness is proportional to protein foldability, as in~\cite{Taverna2002Why,Bloom2006Protein,Zeldovich2007First,Lobkovsky2010Universal}, whereas the opposite case of large $w$ recapitulates flux based energetic models, such as~\cite{Akashi2002Metabolic,Krick2014Amino,Kepp2014Model}. In the following, $w$ will be denoted as chaperone--adjusted synthesis cost.
 


To simulate proteomes evolved to minimize the costs of protein homeostasis, we \hl{used} a lattice protein design, with the following scoring function \hl{for} sequence $S$ at temperature $T$:
\begin{equation}
	\label{score_proteostasis}
	\Pi(S,T,w) = P_{nat} - w\cdot\sum\limits_{a=1}^{20}C_{a}n_{a}
\end{equation}
where $w$ reflects the balance between the costs of amino acid maintenance and chaperone operation, as discussed above. As the exact values of $F,D_{F},U$ and $D_{U}$ \hl{are} not known, we \hl{treated} $w$ as an adjustable parameter, and \hl{optimized} it to find the best agreement between predicted and real amino acid frequencies. Optimization of the proteostasis costs $\mathcal{E}_{p}$ is equivalent to the maximization of the score $\Pi(S,T,w)$. Our protein design procedure \hl{starts} with $M=10^{4}$ random sequences, evaluates the score $\Pi$ for each of them, and evaluates the design criterion for all sequences $S_i$,
\begin{equation}
	\label{design_criterion}
	\frac{1}{M}\sum\limits_{i=1}^{M}\Pi(S_{i},T,w) > \Pi^*
\end{equation}
where $\Pi^*=0.7$ is the threshold proteostasis score defining a viable organism. In the limiting case of $w=0$, the $\Pi^*$ is similar to the parameters previously used in~\cite{Zeldovich2007First}. 

If the criterion \eqref{design_criterion} \hl{is} not satisfied, a random single amino acid mutation \hl{is} introduced in every sequence, and the criterion is reevaluated. The iterative procedure proceeds either until the criterion is met or the number of iterations exceeds $10^{3}$. As the simulation is computationally demanding, GPU accelerated lattice protein folding library GaleProt~\cite{Venev2015Massively} was used for massively parallel evaluation of $P_{nat}$. Therefore, for each combination of the two input parameters, environmental temperature $T$ and chaperone-adjusted synthesis cost $w$, we were able to generate simulated proteomes of $10^4$ sequences of length 64 each, optimizing the proteostasis costs~(\ref{score_proteostasis}). Frequencies of amino acids found in the simulated proteomes were used for comparison with bioinformatics data.

\subsection*{Datasets}
We used RefSeq and BioProject databases at NCBI to retrieve 543 completely sequenced, annotated, single-chromosome bacterial genomes with known optimum growth temperature (OGT) or a specified environmental temperature. A Python script~\cite{Cock2009Biopython} was used to retrieve OGT data from NCBI Entrez. If only a temperature range was specified, the average temperature was used as OGT. The bacterial dataset covers the OGT range of 15--90\textcelsius\, and genome--wide GC content (GC) of 30--70\%, \nameref{fig:s1}.
As archaea are much less represented in the BioProject database, we performed a manual literature search for OGT of 617 species of archaea available in GenBank. The search yielded 223 species with known OGT and sufficient annotation (whole genome shotgun assemblies were included, if at least \PROTEINLIMIT\ protein coding sequences were annotated). Genomes of 83 halophiles have been excluded from analysis~\nameref{table:s1}, as they experience a strong evolutionary pressure of hypersaline environment~\cite{Fukuchi2003Unique}, and appear as outliers on the overall monotonous OGT trends of amino acid usage,~\nameref{fig:s2}. The scatter plots in genomic GC--OGT coordinates for archaea,~\nameref{fig:s1}, reveal a relatively homogeneous coverage, with the GC range 30--70\% and OGT 25--110\textcelsius\ with a lower coverage at $\sim$60\textcelsius\ OGT, which may be attributed to the lack of corresponding environments. The complete dataset comprised 543 bacteria and 140 archaea with sufficient annotation and known OGT, see~\nameref{table:s1} for archaea and~\nameref{table:s2} for bacteria. Analysis scripts and protocols are available at~\url{http://github.com/sergpolly/Thermal_adapt_scripts}.

\subsection*{Identification of highly abundant proteins}
Protein abundance and expression level are important factors to consider when calculating energetic costs. Unfortunately, for most of prokaryotes with completely sequenced genomes neither protein abundance nor expression have been directly characterized, e.g. there are only 2 archaeal entries in the major protein abundance database, PaxDB~\cite{Wang2015Version}. We used a sequence based approach to identify putatively highly expressed proteins using codon adaptation index (CAI)~\cite{Sharp1987The}. Ribosomal proteins were used as a reference of highly expressed proteins~\cite{Pedersen1978Patterns,Srivastava1990Mechanism} to establish the codon usage pattern. We selected all species with at least 25 annotated ribosomal proteins, and used CAI as a proxy for expression and abundance level. Genes with fuzzy locations in the genome have been aligned with the provided protein translation to identify codons in use. Standard BioPython routines were used to calculate CAI~\cite{Cock2009Biopython}. 

Previously, it has been shown that CAI has its limitations as a predictor of gene expression~\cite{Botzman2011Variation}, as in some species the CAI distribution is very narrow and codon usage of ribosomal protein genes is nearly indistinguishable from other genes,~\nameref{fig:s3}. In these cases, the predictive power of CAI is doubtful, as there is no obvious selection for codon usage. To address this issue, we selected a group of genomes where at least 85\% of ribosomal protein genes are within the 25\% of all genes with the highest CAI rank. This empirical criterion selects genomes with wide distributions of CAI and a marked difference in codon usage between ribosomal and other proteins, see~\nameref{fig:s3}, which in turn implies strong codon usage selection (CUS). We assume that in organisms with CUS, CAI can be used as a proxy for gene expression and, statistically, abundance~\cite{Sharp1987The,Jansen2003Revisiting,Supek2005Comparison,Maier2009Correlation}. CUS was identified in about 50\% of species used in this study, 347 bacteria out of 543 and 65 archaea out of 140. Our CUS criterion is compatible with the criteria proposed by Botzman and Margalit~\cite{Botzman2011Variation}, see~\nameref{fig:s4}, and preserves a relatively uniform GC--OGT distribution of species,~\nameref{fig:s1}.

Organisms with CUS were further used to calculate and compare temperature trends of individual amino acids between bacteria and archaea. Highly expressed genes (abundant proteins) were identified for CUS organisms as the genes within the top 10\% of CAI values. Thus we avoided using CAI--ranking for individual genes, which in turn mitigates the problem of poor expression vs abundance correlation~\cite{Maier2009Correlation}.


% Results and Discussion can be combined.
\section*{Results}

\subsection*{Thermal adaptation in highly abundant proteins is similar in bacteria and archaea}

Although archaea and bacteria have diverged early on during evolution, today they share many of the same environments, with both domains spanning wide temperature ranges. Thermal adaptations in the two domains provide a unique test case for comparing phylogenetically divergent responses to the same physical environment. To quantify thermal adaptation, we performed linear regressions between the frequencies of each of the 20 amino acids in the archaeal and bacterial proteomes, and OGT, and used the slopes of the regression as metric of adaptation,~\nameref{fig:s2}. Amino acids with positive slopes are statistically overrepresented in thermophilic proteomes, while negative slopes reflect reduced usage of an amino acid in thermophiles. In a coarse--grained representation, both archaea and bacteria show increased usage of hydrophobic (\texttt{LVIMWPCF}) and charged (\texttt{DEKR}) residues at elevated temperatures, with the corresponding decrease of the polar residues (\texttt{AGNQSTHY}), see~\nameref{fig:s2}. However, at the level of individual amino acids, the correlation between the temperature trends in bacteria and archaea is not statistically significant, Fig.~\ref{fig:fig2}(A) R=0.32 p=0.16, and moreover, bacterial slopes are generally lower than archaeal ones, so the linear regression is shallower than the direct $y=x$ relationship. Therefore, phylogenetic divergence and ensuing biochemical differences had a profound effect on proteome--averaged amino acid usage in the two prokaryotic domains. 

\hl{[Distracting for reader: MOVE ELSEWHERE OR DELETE...]} At the same time, adaptation within phylogenetically close archaeal subclass of {\it Methanococci} bears significant similarities with average proteome--wide archaeal trends of adaptation as shown in~\cite{Haney1999Thermal}, and supported by our calculations (data not shown).

\begin{figure}[h!]
% \makebox[\textwidth][r]{\includegraphics[width=0.9\paperwidth]{../figure2.png}}
\includegraphics[width=\textwidth]{Fig2.png}
\caption{
{\bf Convergence of the archaeal and bacterial trends of thermal adaptation.} Slopes of the amino acid frequency vs OGT regressions are compared between archaeal and bacterial domains.
(A) Proteome-wide, the trends of amino acid usage in bacteria and archaea are not significantly correlated.
(B) ribosomal proteins are used to calculate the trends using all organisms,
(C) proteome-wide trends compared using the species with CUS,
(D) predicted highly expressed proteins (top 10 \% CAI) in the organisms with CUS show identical patterns of thermal adaptation between bacteria and archaea.
}
\label{fig:fig2}
\end{figure}

To look for the common statistical patterns of thermal adaptation between bacteria and archaea, we focused on highly expressed proteins, identified computationally using CAI (see Methods). Highly expressed proteins are known to evolve slowly~\cite{Pal2001Highly,Rocha2004An}, suggesting a stronger evolutionary constraint, which is partially reflected in more stringent folding requirements~\cite{Serohijos2012Protein,Drummond2005Why,Drummond2008MistranslationInduced}. %also known to avoid misinteraction~\cite{Yang2012Protein}. 
\hl{In our model, the selective constraint can be traced to the energy costs of proteostasis being proportional to protein expression levels.} Therefore, we hypothesized that highly expressed proteins experience similar physicochemical selective pressures in archaea and bacteria, so their thermal adaptation mechanisms may \hl{have converged} despite the deep phylogenetic divergence between domains; in other words, for highly expressed proteins, the physical and energetic evolutionary constraints may prevail over phylogenetic history.

Ribosomal proteins serve as a particularly well--defined group of highly expressed proteins in both archaea and bacteria~\cite{Karlin2005Predicted}. At the same time, differences in ribosome structures and sequences between the two domains are significant. Remarkably, both domains of life exhibit very similar strategies in thermal adaptation of ribosomal proteins, Fig.~\ref{fig:fig2}(B), R=0.73, p\textless 0.001 (bootstrap to find a similar correlation in the same number of randomly selected proteins yields p\textless0.001, see~\nameref{text:s1} for details). Additionally, the observed relation between archaeal and bacterial slopes is close to equality ($y=x$), making this effect quantitative. However, the specific function of ribosomal proteins may have limited their options for thermal adaptation. We used organisms with CUS to identify other types of highly expressed proteins, and repeated the analysis \hl{to test if similarity in thermal adaptation is inherent to broader class of highly expressed proteins.}

Remarkably, for predicted highly expressed proteins from organisms with CUS, the trends in thermal adaptation are \hl{indeed} nearly identical ($y=x$). Fig.~\ref{fig:fig2}(D) demonstrates R=0.87, p\textless 0.001 for all proteins within the top 10\% of CAI, while excluding ribosomal proteins yields R=0.873 (data not shown). Null hypothesis that the observed correlation can be \hl{explained by a random draw} is safely rejected (p=0.026, randomized CUS selection,  p\textless0.001, randomized CAI ranking, see~\nameref{text:s1} for details).

\hl{It is important to note that CUS alone does not warrant similarity in thermal adaptation for bacteria and archaea.} Trends in thermal adaptation in complete proteomes of bacteria and archaea with CUS appear similar, Fig.~\ref{fig:fig2}(C), R=0.55, p=0.01. However, R=0.55 can be achieved with p$\approx$0.3 using randomized selection of \hl{the same number} of bacteria and archaea from the full dataset, see~\nameref{text:s1} for details. Thus, we cannot reject the null hypothesis \hl{that CUS does not affect this type of correlation}. Moreover, both proteome--wide trends from Fig.~\ref{fig:fig2}(A) and (C) deviate substantially from the equal relationship ($y=x$), emphasizing importance of phylogenetic factors on the adaptation of entire proteomes from different domains. 

Therefore, bioinformatics analysis suggests that highly expressed proteins in both archaea and bacteria share a common strategy of thermal adaptation, which becomes obscured at the level of complete proteomes. We propose that the common strategy may involve optimization of energetic costs of proteostasis by balancing amino acid metabolism and chaperone energy expenses, and present the results of the modeling below. 

\subsection*{Simulated environmental temperature affects amino acid composition}
\subsection*{\hl{Amino acid usage in simulated proteomes}}

We designed lattice model proteins in a wide range of artificial temperatures $0.4\leq T\leq 1.7$ in units of Miyazawa--Jernigan residue level potential (p.u.)~\cite{Miyazawa1999SelfConsistent}. The chaperone--adjusted synthesis cost was varied from $w=0$, implying no cost of amino acid maintenance, to $w=0.15$, where, \hl{on the contrary, costs of amino acid maintenance} $C_{a}$ governed amino acid frequencies in simulated proteomes \hl{with little effect of folding requirement}. Proteins designed with no synthesis costs constraint, $w=0$, mostly reproduced earlier results~\cite{Berezovsky2007Positive,Venev2015Massively}. At low simulated temperatures, the folding constraint on protein sequences was weak. Starting from a random sequence with $\approx 1/20$ amino acid abundances, the design procedure was able to create well--folding sequences by swapping the residues while retaining the overall amino acid composition, Fig.~\ref{fig:fig3}. As the temperature increased, relative amino acid abundances changed monotonically to allow designed proteins to increase their thermal stability, Fig.~\ref{fig:fig3}(A, inset). As shown earlier~\cite{Berezovsky2007Positive}, increasing frequencies of hydrophobic and charged residues extend the energy gap by decreasing the energy of the native state and increasing the average decoy energy. 

\begin{figure}[h!]
% \makebox[\textwidth][r]{\includegraphics[width=0.9\paperwidth]{Figure_3.png}}
\includegraphics[width=\textwidth]{Fig3.png}
\caption{
{\bf Amino acid frequencies in the simulated proteomes change with temperature} (A) Simulated amino acid frequencies, neglecting the metabolic costs of amino acid synthesis, $w=0$.  (B) Metabolic costs constraint on the simulated proteomes, $w=0.06$, alters both low-temperature distribution of amino acid usage and their overall temperature trends. (inset) Simulated trends for charged, hydrophobic and hydrophilic groups of amino acids match experimentally observed trends~\cite{Berezovsky2007Positive}.
}
\label{fig:fig3}
\end{figure}

Although the temperature trends of amino acid groups in simulated proteomes are similar to natural ones~\cite{Berezovsky2007Positive,Venev2015Massively}, the simulated frequencies of individual amino acids do not match bioinformatics data (typical $R\approx0.3$ for $0.4\leq T\leq 1.7$ p.u., see Fig.~\ref{fig:fig4}(B), $R_A$ profile at $w=0.0$). We hypothesized that by introducing the proteostasis cost, it will be possible to design protein sequences where both the average composition and its temperature trends are reflective of reality. Indeed, the outcome of protein design changed significantly if the chaperone--adjusted synthesis cost $w$ increased, Fig.~\ref{fig:fig3}(B). In this case, frequent usage of ``expensive'' amino acids carried a significant penalty even if they were favorable for protein foldability. At $w=0.06$, proteome--averaged frequencies of amino acids already diverged at low temperatures $T$, and the distribution of amino acid frequencies was mostly determined by their relative metabolic maintenance costs, due to the smaller selective pressure on foldability. The average frequencies of all amino acids varied strongly according to their metabolic costs, unlike in Fig.~\ref{fig:fig3}(A), where well--folding sequences at $w=0$ could be made by small changes in amino acid composition. However, even at $w>0$, the temperature trends of groups of amino acids remained generally similar to the case of $w=0$, Fig.~\ref{fig:fig3}(B, inset).


\begin{figure}[h!]
% \makebox[\textwidth][r]{\includegraphics[width=0.9\paperwidth]{../exp_MTAD_all_all_bact_summary_Figure_4.png}}
\includegraphics[width=\textwidth]{Fig4.png}
\caption{
{\bf Simulated frequencies of amino acids compared with the naturally evolved ones for bacteria.} 
Pearson correlation coefficient $R_M, R_T$ between simulated and observed frequencies in mesophilic and thermophilic bacteria, respectively, as function of the simulated temperature and chaperone-adjusted synthesis cost $w$. Proteome-averaged amino acid frequencies in bacteria were used. The best correlation between simulated and experimental data is achieved for $w=0.07$ for thermophiles (A), and $w=0.06$ for mesophiles (B). For the combination of mesophilic and thermophilic species (C), the overall best correlation $R_A$ between simulated and observed data is observed for $1<T<1.2$ (highlighted in yellow). The same temperature range yields a high correlation $R_D$ between the simulated and observed temperature trends of amino acid frequencies $df/dT$, (D).
}
\label{fig:fig4}
\end{figure}

% may need to join the the sections ...
\subsection*{Simulated trends correlate with bioinformatics data}

As shown in Fig.~\ref{fig:fig3}, amino acid frequencies produced by our model are controlled by two parameters, temperature $T$, and the chaperone-adjusted synthesis cost $w$. We hypothesized that natural amino acid frequencies and their temperature trends could be well reproduced simultaneously by an appropriate choice of $w$ and temperature range. We used the Pearson correlation coefficient between naturally evolved and simulated frequencies of amino acids to assess the fit of the model to observed frequencies for given $w$ and $T$. Prokaryotic genomes were separated into mesophilic (20$\leq$OGT$\leq$ 50\textcelsius) and thermophilic (OGT$\geq$ 50\textcelsius) groups and average amino acid frequencies from these groups were used for comparison with the simulated data using correlation coefficients $R_M(T,w)$ and $R_T(T,w)$ respectively. This analysis has been performed separately for bacteria and archaea. In bacteria, the correlation coefficients $R_M$ and $R_T$ behave smoothly, and reach their maxima at specific values of $T$ and $w$, Fig.~\ref{fig:fig4}. The absolute values of $R_M$ and $R_T$ were extremely high ($R\approx0.9$), similar to the phenomenological model~\cite{Krick2014Amino}. 

Our model correctly segregated thermophilic and mesophilic genomes. The value of $R_M$ reached its maximum at $T_M=0.9$ p.u. while $R_T$ reached the maximum at a higher temperature $T_T=1.2$, Fig.~\ref{fig:fig4}(A,C). Both $R_M$ and $R_T$ reached their maxima at similar values of chaperone--adjusted synthesis cost $w_M\approx w_T \approx 0.06$, so the energetic balance between chaperone activity and costs of amino acid maintenance appeared similar between thermophiles and mesophiles.

To find the best global fit, or the range of parameters where our model best represents both mesophilic and thermophilic proteomes, we used $R_M + R_T$ as the metric describing the overall agreement between model predictions and bioinformatics observations on amino acid frequencies and their temperature trends. To avoid numerical instabilities when finding the maximum of $R_M+R_T$ with respect to $(T_M, T_T, w)$, we first established the optimum value of $w$ from
\begin{equation}
	\label{optimal_w_definition}
	w^* = \argmax_{w} ( R_M(T_M, w) + R_T(T_T,w) ),
\end{equation}
where $T_M=\argmax_{T}R_M(T,w), \quad T_T=\argmax_{T}R_T(T,w))$, i.e. we maximized the correlations with respect to $T$ separately for mesophiles and thermophiles first, see Fig.~\ref{fig:fig4}(A,C), and then optimized $w$ using the combined metric. To check for consistency of this procedure, we have then used $w^*$ and found the simulated temperatures best fitting mesophiles and thermophiles, 
\begin{equation}
	\label{optimal_T_range}
	T^*_M = \argmax_{T}R_M(T, w^*), \quad T^*_T = \argmax_{T}R_T(T, w^*).
\end{equation}
Specifically, we found $w^*=0.06,\  T^*_M=0.9,\ T^*_T=1.1$ p.u., see \nameref{fig:s6}. Therefore, our procedure finds a self-consistent set of parameters describing the temperature range between mesophiles and thermophiles, and the adjusted synthesis cost $w$. These parameters successfully describe the complete dataset of both mesophiles and thermophiles in terms of amino acid composition and its temperature trends. Fig.~\ref{fig:fig4}(B) shows the fit between predicted amino acid frequencies $\vec f_{model}$ and the full set of all genomes, measured as the correlation coefficient $R_A = R_{M \cup T}(\vec f_{model}, \vec f_{exp})$. The maximum correlation of $R=0.93, p<0.001$ is highly statistically significant. Importantly, the same set of model parameters describes well the temperature trends of amino acid composition, or slopes $df_i/dT$ for most amino acids. Fig.~\ref{fig:fig4}(D) shows the correlation $R_D = R_{M\cup T}(d\vec f_{model}{/dT}, d\vec f_{exp}/dT)$. Similar to $R_A$, the value of $R_D$ exhibits a clear maximum with respect to both $w$ and $T$, reaching $R_D\approx0.60$, similar to earlier findings~\cite{Venev2015Massively}.

The best fit of the model to the experimental data was achieved for $1.0<T<1.2$ p.u. and $w=0.07$; for those parameters, $R_A=0.93$ and $R_D=0.60$, both very highly statistically significant values. Interestingly, the relative temperature range in the model, $(T_T-T_M)/T_M\approx 20\%$ compared well with the actual temperature range of prokaryotes, thriving between approximately 280$K$ and 370$K$, a 30\% change in absolute temperature. Comparison between simulated data and amino acid frequencies in archaea is presented in~\nameref{fig:s5}, and shows similar values of the optimum temperature range and chaperone-adjusted synthesis cost $w$.

To prove that these results are not a numerical artifact, we have shuffled the values of amino acid maintenance cost $C_a$ and repeated the simulations. The reshuffling breaks the connection between the biochemistry of amino acids, reflected in their costs $C_a$, and their physical properties, such as interaction energies $\epsilon_{ij}$. As described in~\nameref{text:s2}, the values of $R_A$ and $R_D$ obtained from reshuffled $C_a$ are almost always lower than those for the initial true $C_a$, yielding $p<0.001$,~\nameref{text:s2}. Therefore, we demonstrated that the interplay between amino acid synthesis costs and protein folding is internally consistent, and our model produces correct amino acid frequencies only when the connection between the biochemical and physical properties of amino acids is preserved.

 
\subsection*{Predicted temperature trends of specific amino acids}
    
We used difference quotients of simulated amino acid frequencies $\Delta\mathit{f}_{a}/\Delta T, a=1\dots20$ in the $[T^*_M,T^*_T]$ range to calculate the simulated temperature trends and compared them to bioinformatics data. For the natural frequencies of amino acids slopes were derived from the linear regression analysis over the entire OGT range, see~\nameref{fig:s2}.

The complete proteomes of both bacteria and archaea produced similar, statistically significant correlations with model predictions, $R=0.59$ and $R=0.61$, respectively, Fig.~\ref{fig:fig5}(A,B). We have then considered only highly expressed proteins (top 10\% of CAI for organisms with CUS) from either domain, expecting that the selective pressure of proteostasis is stronger for this group of proteins. However, we did not find a significant difference between the temperature trends in complete proteomes vs highly expressed proteins, Fig.~\ref{fig:fig5}(C,D). Consistent with previous findings~\cite{Venev2015Massively}, the temperature trend of leucine (\texttt{L}) frequency was not captured well by the model. Leucine is a very hydrophobic residue, as reflected by the Miyazawa and Jernigan interaction potential. Accordingly, the frequency of leucine rapidly increased with temperature in simulated proteomes, as leucine participates in hydrophobic interactions in the protein core. However, the frequency of leucine does not increase with temperature in bacteria, although it does so in archaea, see~\nameref{fig:s2}. Coupled to the fact that leucine is relatively simple to synthesize, and is coded by 6 different codons, these observations clearly point to the biochemical differences between archaea and bacteria, and to the limitations of current biophysical models. Aspartic acid (\texttt{D}) is an another outlier. This charged amino acid is predicted to increase in frequency as the temperature rises, just as glutamic acid, lysine, and arginine (\texttt{E}, \texttt{K}, \texttt{R}). However, while glutamic acid and lysine consistently increase in frequency in both bacteria and archaea, aspartic acid is surprisingly depleted in natural thermophilic proteomes.

\begin{figure}[h!]
% \makebox[\textwidth][r]{\includegraphics[width=0.9\paperwidth]{Figure5.png}}
\includegraphics[width=\textwidth]{Fig5.png}
\caption{
{\bf Simulated trends of thermal adaptation in amino acid compositions are compared with the observed one.} Simulated slopes vs the evolved slopes are presented for all 20 amino acids using proteome--wide data from (A) archaea and (B) bacteria, and using top 10\% of CAI ranked proteins in the organisms with CUS only, both for (C) archaea and (D) bacteria. Slopes predicted from the simulations are in good agreement with the naturally evolved ones, correlation coefficients $R=0.56$ to $R=0.61$, $p<0.01$. The frequency of leucine is consistently overestimated by the simulations.
}
% \label{fig:aa_slopes}
\label{fig:fig5}
\end{figure}


\subsection*{Amino acid synthesis costs increase with environmental temperature}

Metabolic costs of amino acid synthesis are negatively correlated with protein expression levels across the three domains of life~\cite{Akashi2002Metabolic,Swire2007Selection}. In Fig.~\ref{fig:fig6}(A,C), we plotted the proteome--averaged Akashi--Gojobori amino acid synthesis cost against environmental temperature for 140 archaea and 543 bacteria, assuming equal expression levels of all proteins. We found a statistically significant positive correlation, confirming that thermal stability requires heavier usage of synthetically ``expensive'' proteins, in agreement with an earlier observation made on {\it Thermus thermophilus} genome~\cite{Swire2007Selection}. In contrast with the amino acid synthesis cost, the amino acid maintenance cost, which combines synthesis and decay~\cite{Krick2014Amino}, is not significantly correlated with the environmental temperature (OGT) Fig.~\ref{fig:fig6}(B,D). These observations are fully reproduced by our model, Fig.~\ref{fig:fig6}(E,F), for the case of bacteria: in the relevant temperature range $1.0 < T < 1.2$, p.u. and $w^*=0.07$, the synthesis cost of evolved proteomes increased with temperature, while the average amino acid maintenance cost did not. For archaea, the predicted temperature range was wider due to numerical instabilities in finding the optimum values of $w$ and $T$, see~\nameref{fig:s7} for details.


\begin{figure}[h!]
% \makebox[\textwidth][r]{\includegraphics[width=0.99\textwidth]{Figure7.png}}
\includegraphics[width=\textwidth]{Fig6.png}
\caption{
{\bf Temperature trends of the amino acid synthesis and maintenance costs in natural and simulated proteomes.}
Proteome average cost of amino acid synthesis and amino acid maintenance for archaeal species (A,B) and bacterial species (C,D). Marker color reflects the genome--wide GC content of each specie. Linear regression is provided for each of the trends.
Simulated trends are presented for bacteria in a wide range of chaperone--adjusted synthesis cost $w$ (E,F), including the overall optimal one $w^*=0.07$ (green squares). Yellow-shaded area highlights optimal temperature span that is best correlated with the observed data. Costs trends within the highlighted $T$-range reproduce observed trends in bacteria.
}
\label{fig:fig6}
\end{figure}

\subsection*{Highly expressed proteins are similar to thermophilic ones in bacteria, not in archaea}

In full agreement with earlier findings by Akashi and Gojobori~\cite{Akashi2002Metabolic} and Swire~\cite{Swire2007Selection}, our analysis found a statistically significant negative correlation between the amino acids synthesis costs and CAI (proxy for expression) in sets of 347 bacterial and 65 archaeal genomes, see~\nameref{fig:s8}. At the same time, it has been proposed that amino acid composition of highly expressed proteins is similar to the composition of thermophilic proteins~\cite{Cherry2010Highly}. As noticed earlier~\cite{Serohijos2012Protein}, this is somewhat contradictory to the findings of Akashi et al and Swire, as thermophilic proteins tend to be more synthetically ``expensive''.

To look for the origins of this controversy, we compared the protein expression levels, approximated by CAI, with their thermostability in bacteria and archaea. To assess protein thermostability of a group of proteins, we used the Pearson correlation coefficient $R_T$ between their average amino acid compositions \hl{and} 92 thermophilic archaea \hl{or} 66 thermophilic bacteria (OGT\textgreater50\textcelsius). \hl{For each organism, proteins were split into five groups, corresponding to the quintiles of their CAI, and average amino acid usage of each quintile was compared with the thermophilic composition using $R_T$.} Bacteria exhibited a statistically significant positive correlation between the mean $R_T$ and CAI quintile \hl{(expression proxy)}, Fig.~\ref{fig:fig7}(B). As a control, we have reshuffled synonymous codons within each genome, completely destroying the codon bias and thus CAI metric of each protein, but leaving amino acid composition intact. No correlation was observed in reshuffled data for bacteria, see Fig.~\ref{fig:fig7}(B). These results support Cherry's finding of highly expressed proteins having amino acid composition similar to thermophilic ones. However, in the case of archaea, we did not observe a correlation between CAI and $R_T$, see Fig.~\ref{fig:fig7}(A). Moreover, for archaea, synonymous codon reshuffling resulted in a strong negative correlation between CAI and $R_T$, see Fig.~\ref{fig:fig7}(A). These results partially support Cherry's findings and demonstrate the immense flexibility of the 20--dimensional space of protein sequence composition to satisfy multiple physical and phylogenetic constraints.


\begin{figure}[h!]
% \makebox[\textwidth][r]{\includegraphics[width=0.9\paperwidth]{Figure7_quantiles.png}}
\includegraphics[width=\textwidth]{Fig7.png}
\caption{
{\bf Similarity between proteins with high CAI and thermophilic proteins by comparing amino acid composition.}
Proteins in each organism are grouped into 5 quantiles according to their CAI value, and amino acid composition within these groups is compared with the averaged thermophilic composition using Pearson correlation coefficient $R_T$, separately for archaea (A), and bacteria (B). In bacteria, the higher is protein expression, the more similar is amino acid composition to a thermophilic one (increasing $R_T$, red line, statistically significant positive correlation, p=0.005). No such trend was observed for archaea. As a control, codon reshuffling was used to destroy the relation between CAI and amino acid composition of proteins. For both archaea and bacteria, the correlation between CAI and $R_T$ for reshuffled codons was not significant.  Error bars represent the 30\% and 70\% percentiles of the underlying distrbutions,~\nameref{fig:s9}.
}
\label{fig:fig7}
\end{figure}

We have also tried to approximate protein thermostability by the fraction of \texttt{IVYWREL} amino acids, which is strongly correlated with OGT at the proteomic level~\cite{Zeldovich2007Protein}. Although a strong negative correlation between \texttt{IVYWREL} and CAI was observed, see~\nameref{fig:s10}, the same trend persisted upon synonymous codon reshuffling, in both bacteria and archaea. Therefore, an intrinsic connection between the amino acid and nucleotide frequencies via the genetic code precludes use of \texttt{IVYWREL} metric for comparing protein expression (CAI) and thermostability.

\section*{Discussion}


Statistically significant correlations between environments and amino acid usage are well established, dating back at least to 1986, when Ponnuswamy et al hypothesized that amino acid usage can be quantitatively linked to environmental temperature~\cite{Ponnuswamy1986Amino}. Subsequent studies of complete genomes demonstrated that variation in amino acid composition can be attributed largely to two independent factors,  GC content and environmental  temperature~\cite{Kreil2001Identification,Singer2003Thermophilic}. Understanding the microscopic origins of these correlations requires both statistical analysis and modeling of the impact of amino acid compositions on organismal fitness. 

An early model of amino acid composition has been proposed by Dill~\cite{Dill1985Theory}, who derived the ratio of hydrophobic to polar residues conferring highest stability to a globular protein. The interest in the statistical understanding of thermal adaptation increased as microscopic simulations of protein evolution became possible~\cite{Taverna2002Why,Bloom2006Protein,Goldstein2008The}. Berezovsky et al~\cite{Berezovsky2007Positive} hypothesized that protein foldability was the main selective pressure responsible for thermal adaptation, and analyzed amino acids usage in 27-mer lattice proteins designed to be stable in a wide range of temperatures. Although the temperature trends in amino acid frequencies could be explained by a purely physical model, the frequencies themselves were weakly correlated with genomic data. Extension of the folding model to 64-mer lattice proteins yielded only a marginal improvement~\cite{Venev2015Massively}. This continued discrepancy suggests that either the physical models are still not precise enough to resolve individual amino acid  beyond their rough classification by hydrophobicity, or other factors, not directly related to protein folding, control amino acid usage.

Complementary to protein folding constraints, metabolic costs and overall energy balance of a cell have been long identified as powerful evolutionary drivers~\cite{Pal2006An}, as exemplified e.g. by the success of quantitative flux based metabolic models~\cite{Varma1994Metabolic,Price2004Genome}. Akashi and Gojobori estimated the energy expended on the synthesis of each of the 20 types of amino acid molecules, and found that highly expressed proteins are enriched in ``cheap'', easily synthesized amino acids~\cite{Akashi2002Metabolic}. These findings highlighted the importance of proteostasis as the major cellular process, coupling energy and material fluxes in a cell. The flux models were further advanced by an estimate of the amino acid decay rates within a cell~\cite{Krick2014Amino}. By combining the amino acid synthesis cost, decay rate, and sequence entropy into an empiric cost function, Krick et al made successful predictions of amino acid frequencies~\cite{Krick2014Amino}. However, this model did not explicitly address protein folding or other physical considerations, and so it is difficult to extend it to the study of thermal adaptation.

To bridge this gap, we proposed that proteostasis is not limited to the chemical turnover of amino acid molecules, but, crucially, maintains appropriate levels of functional, correctly folded proteins. Molecular chaperones are an integral part of this process, attempting to refold proteins in an ATP-dependent manner. Invoking quality control systems in response to misfolded proteins causes a fitness penalty proportional to the fraction of misfolded proteins, their expression level and is largely function--independent~\cite{Samerotte2011Misfolded}. Moreover, further experiments suggested that it is indeed the metabolic cost of chaperone activation and action that imposes the fitness penalty, rather than the consequences, e.g. toxicity, of the presence of abundant misfolded proteins~\cite{Tomala2014Fitness}. Chaperone function provides a feedback to the genotype, by acclereating its evolution while serving as a capacitor for otherwise deletirious phenotipic mutations~\cite{Bogumil2012Cumulative,Cetinbas2013Catalysis}

Following Kepp et al~\cite{Kepp2014Model}, we hypothesized that the energy consumed by chaperones is non-negligible and must be taken into account together with other metabolic costs. Specifically, we assumed that the total energy cost of proteostasis includes contributions from both amino acid turnover and chaperone activity. The key feature of the model is the statistical dependence between foldability of a protein and its amino acid composition~\cite{Dill1985Theory,Berezovsky2007Positive,Venev2015Massively}. Indeed, well--folded proteins typically contain a balanced mix of charged and hydrophobic residues, while intrinsically unfolded proteins do not~\cite{Uversky2000Why}. Proteins with an imbalanced amino acid composition, statistically, are less stable and so may require more frequent chaperone intervention. Therefore, we posited that amino acid compositions have evolved to minimize the total energy spent on amino acid homeostasis and chaperone activity, and tested this hypothesis by simulations.

By incorporating protein folding and metabolic cost in a single model, we were able to capture average amino acid composition and its temperature trends simultaneously, Fig.~\ref{fig:fig4}, significantly improving upon purely physical models~\cite{Berezovsky2007Positive,Venev2015Massively}. These  models are captured in our study as a limiting case $w=0$. As demonstrated in Fig.~\ref{fig:fig4}, the predictive power of the model dramatically increases by considering an interplay between protein folding requirement and the metabolic cost constraints, $w\neq 0$. Although the protein folding requirement has been long associated with fitness~\cite{Taverna2002Why,Bloom2006Protein,Zeldovich2007First,Lobkovsky2010Universal}, our present model provides an explicit link between folding defects and organismal fitness via elevated cost of chaperone activity.

It is instructive to consider the temperature trends of estimated cost of mesophilic and thermophilic proteomes according to two different metrics, amino acid synthesis cost as defined by Akashi and Gojobori~\cite{Akashi2002Metabolic}, and the amino acids maintenance costs derived by Krick et al~\cite{Krick2014Amino}. As shown in Fig.~\ref{fig:fig6}(A,C), the Akashi--Gojobori metabolic cost markedly increases in thermophilic proteomes, p\textless0.001 both for archaea and bacteria. This finding is explained by the lower costs of small, polar amino acids according to this scale, compared to larger ones, either hydrophobic or charged. However, the amino acid maintenance cost is not significantly correlated with temperature in bacteria and weakly decreases with temperature in archaea, Fig.~\ref{fig:fig6}(B,D). This may be interpreted as the lower fraction of metabolic costs  of proteostasis in the energy budget of thermophilic organisms or by severe energetic constraints imposed on thermophilic organisms. At the same time, the mathematical origins of this result are evident from the comparison of the respective amino acid costs, as contrary to the Akashi--Gojobori cost $S_{a}$, hydrophobic, charged and polar classes of amino acids are widely scattered in terms of the degradation--corrected Krick cost. Overall these additional observations support the results of our simulations and emphasize the importance of the selective pressure acting on protein homeostasis during evolution. Previously, the strong influence of protein homeostasis costs on the evolution has been shown to be as strong, if not stronger, than selection for protein function~\cite{Assis2014Conserved}.

Metabolic synthesis costs are likely proportional to protein abundance, and so are correlated with expression levels, which can be either comprehensively measured using RNAseq or other techniques, or inferred from genomic data. It is well known that protein synthesis costs are negatively correlated with expression levels in multiple organisms~\cite{Akashi2002Metabolic,Seligmann2003CostMinimization,Heizer2006Amino,Raiford2008Do}. Similarly, it has been shown that highly expressed proteins experience lower rates of evolution~\cite{Pal2001Highly,Rocha2004An,Drummond2005Why}, and so presumably undergo stronger selection. The mutual dependence of protein expression levels, stability, and evolutionary rate has been addressed by recent biophysical modeling~\cite{Serohijos2012Protein}.

Our analysis indicates that thermophilic proteins have an increased cost according to the Akashi--Gojobori metric, Fig.~\ref{fig:fig6}(A,C), while highly expressed proteins are known to be ``cheap''~\cite{Akashi2002Metabolic}. On the other hand, it has been suggested that amino acid composition of highly expressed proteins is similar to that of thermophilic proteins~\cite{Cherry2010Highly}, creating a logical inconsistency. We attempted to address this issue by estimating the expression levels using CAI and correlating it with various composition--based predictors of thermostability in a large set of bacterial and archaeal proteomes. In the bacterial dataset, we observed that highly expressed proteins had amino acid compositions more similar to the average composition of thermophilic proteomes. This finding parallels earlier results of Cherry~\cite{Cherry2010Highly}. However, no significant correlation was found in archaea, see \nameref{fig:s10}. 

At the same time, the temperature trends in amino acid frequencies of highly expressed proteins in archaea and bacteria are strongly correlated, Fig.~\ref{fig:fig2}(D), while proteome--wide correlation is much lower, Fig.~\ref{fig:fig2}(A). This convergence of thermal adaptations in highly expressed but overall divergent proteins suggests a common selective pressure, such as metabolic or proteostasis costs. Apparent inconsistencies in the cost--expression--stability logic loop require further study, and may evidence a surprising flexibility of amino acid usage evolving to satisfy different constraints. Further development of high-throughput experimental methods for characterizing protein expression levels and stability will make it possible to transition away from sequence--based predictors, and will stimulate the next generation of predictive, organism--level models of metabolism and selection.




% Do NOT remove this, even if you are not including acknowledgments.
\section*{Acknowledgments}
We acknowledge the help of Alexey Shaytan, Alexander Goncearenco, and NCBI helpdesk for assistance with the NCBI databases.


\section*{References}
% Either type in your references using
% \begin{thebibliography}{}
% \bibitem{}
% Text
% \end{thebibliography}
%
% OR
%
% Compile your BiBTeX database using our plos2009.bst
% style file and paste the contents of your .bbl file
% here.
% 

\bibliography{Remote}


\section*{Supporting Information}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% supp texts 
\subsection*{S1 Text}
\label{text:s1}
% \nameref{text:s1}
{\bf Thermal adaptation ``convergence'' tested using statistical bootstraps}
Description of statisticall procedures, including random sampling of organisms to test significance of CUS, and random sampling of proteins to test significance of selected groups of proteins.
% Description of the statisticall procedures to test plausibility of null hypothesis for the adaptation convergence analysis. Selection of ribosomal proteins is bootstrapped selection of 50 random CDS from the proteome and selection of organisms with CUS is tested using selection of \%40 random organisms.


\subsection*{S2 Text}
\label{text:s2}
{\bf Statisticall validation of the simulated amino acid compositions and trends.}
Reshuffling of the amino acid maintenance costs $C_{a}$ is used to validate simulations results, including the correlation between simulated and observed amino acid frequencies $R_A$ and trends $R_D$.



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% supp figures 
\subsection*{S1 Fig}
\label{fig:s1}
{\bf Dataset quality plot in GC--OGT corrdinates.}
Variety of analysed archaeal and bacterial species is demonstrated. \hl{Balanced distrbution in GC--OGT plane allows for fair and unbiased analysis.}


\subsection*{S2 Fig}
\label{fig:s2}
{\bf Thermal adaptation: temperature trends for 20 amino acids. }
Proteome--wide frequencies of 20 amino acids plotted as functions of OGT for archaea and bacteria. Trends of importnat amino acid combinations are demonstrated as well, including \texttt{IVYWREL}, hydrophobic, hydrophilic and chared residue types. Halophilic archaea stand out from smooth transition between mesophils and thermophiles.


\subsection*{S3 Fig}
\label{fig:s3}
{\bf CAI distribution example for organisms with CUS and without it.}
Proteome--wide CAI distribution is demonstrated along with the CAI distribution of ribosomal proteins for a pair of selected organisms. 


\subsection*{S4 Fig}
\label{fig:s4}
{\bf CUS criteria comparison: distribution of mean CAI}
Distribution of the proteome--wide mean CAI is demonstrated for organisms with CUS and without it.
CUS organisms identified by our criteria indeed have higher value mean CAI, as suggested earlier~\cite{Botzman2011Variation}.


\subsection*{S5 Fig}
\label{fig:s5}
{\bf Frequencies of amino acids from simulated proteomes compared with the naturally evolved ones.} Pearson correlation coefficient is used to compare amino acid frequencies for mesophiles, $R_M$, thermophiles, $R_T$, average amino acid composition $R_A$, and with the slopes of amino acid temperature trends, $R_D$. Corresponds to the Fig.~\ref{fig:fig4}, but for archaeal dataset and predicted highly expressed proteins from organisms with CUS.



\subsection*{S6 Fig}
\label{fig:s6}
{\bf Overall optimal chaperone--adjusted synthesis cost determined from $R_M$ and $R_T$.}
Equally weighted dependencies of $R_M$ and $R_T$ on the chaperone-adjusted synthesis cost $w$ are combined to determine its optimal values. Comparison with both archaeal and bacterial data presented.


\subsection*{S7 Fig}
\label{fig:s7}
{\bf Temperature trends of the simulated amino acid synthesis and maintenance costs}
Simulated trends of proteome average cost of amino acid synthesis and maintenance are presented for archaeal dataset and predicted highly expressed proteins from organisms with CUS. Optimal chaperone--adjusted synthesis cost $w^*$ and temperature ranges are highlighted.



\subsection*{S8 Fig}
\label{fig:s8}
{\bf Metabolic synthesis cost trends with CAI.}
Proteins divided into 5 groups according to their CAI value are used to demonstrate statistically significant decrease in biosynthesis costs both for archaea and bacteria.


\subsection*{S9 Fig}
\label{fig:s9}
{\bf Similarity between predicted highly expressed proteins and themophilic proteins: details of distribuitons.}
Proteins divided into 5 quintiles according to their CAI value and quintile--averaged usage of amino acids is compared with thermophilic composition of amino acids using $R_T$. Values of $R_T$ for different organisms compose distributions for each quintile. These distributions correspond to the mean $R_T$ trends, Fig.~\ref{fig:fig4}.



\subsection*{S10 Fig}
\label{fig:s10}
{\bf Proteomic measure of thermostability \texttt{IVYWREL} trends with CAI.}
Proteins divided into 5 groups according to their CAI value and group--averaged fraction of \texttt{IVYWREL} amino acids is calculated. Negative correlation of mean \texttt{IVYWREL} is demonstrated both for original data and for codon--reshuffled control, thus suggesting intrinsic link between \texttt{IVYWREL} and the genetic code.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{S1 Table}
\label{table:s1}
% \nameref{table:s1}
{\bf Table with the archaeal species used for analysis}
Species used for the analysis are listed in the Table along with some basic information.
AssemblyID is used as a unique identifier of species, and a particular FTP--link of the genome assembly is provided as well.
Organism name and related taxonomic IDs are reported in corresponding columns. OGT is reported along with the web-link to the source of OGT information, ranging from peer--reviewed articles to database entries. CUS titled column reports if an organism demonstrate codon usage selection according to our criteria. Halophiles excluded from the analysis, i.e., phylogenetic subdivisions {\it Halobacteria} and {\it Nanohaloarchaea}, are marked in the Halophilic column.



\subsection*{S2 Table}
\label{table:s2}
% \nameref{table:s1}
{\bf Table with the bacterial species used for analysis}
Species used for the analysis are listed in the Table along with some basic information.
Genome accession id is reported in GenomicID column and is used as a unique specie identifier.
Organism name and related taxonomic IDs are reported in corresponding columns.
OGT is reported along with the corresponding id of a BioProject entry that contains the OGT.
CUS titled column reports if an organism demonstrate codon usage selection according to our criteria.






% This work was supported in part by the Defense Advanced Research Projects Agency [D13AP00041 and HR0011-11-C-0095].


% venevs@ummsres25:~/Dropbox (UMASS MED - BIB)/Thermal_adapt_scripts/Publication$ python DONE_get_values_paper.py
% # of archaeal assemblies in the GenBank DB at NCBI, 617
% # of archaeal genomes with sufficient annotation and OGT,  223
% # of archaeal genomes with sufficient annotation and OGT (Halophiles excluded),  140
% # of bacterial genomes with full annotations and OGT (analyzed genomes),  543
% Total # of genomes analysed: archaea without Halophiles, bacteria complete genomes only = 683

% Thermophiles ...
% # of archaeal Thermophiles with sufficient annotation and OGT,  92
% # of bacterial Thermophiles with full annotations and OGT,  66
% CUS stuff ...
% # of Archaeal species with Translational Optimization,  91
% # of Archaeal species with Translational Optimization (excluding Halophiles),  65
% # of Bacterial species with Translational Optimization,  347



% \section*{Tables}
% 
% See introductory notes if you wish to include sideways tables.
%
% NOTE: Please look over our table guidelines at http://www.plosone.org/static/figureGuidelines#tables to make sure that your tables meet our requirements. Certain types of spacing, cell merging, and other formatting tricks may have unintended results and will be returned for revision.
%
%\begin{table}[!ht]
%\caption{
%\bf{Table title}}
%\begin{tabular}{|c|c|c|}
%table information
%\end{tabular}
%\begin{flushleft}Table caption
%\end{flushleft}
%\label{tab:label}
% \end{table}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% \begin{table}[!ht]
% 	\caption{\bf{Temperature trends at amino acid resolution}}
% 	\begin{tabular}{|c|c|c|c|c|}
% 		% \toprule
% 		{} &   {\bf Archaea} &  $\mathbf{GC_{30}}$ &   $\mathbf{GC_{50}}$ &  {\it\bf Methanococci} \\
% 		% \midrule
% 		{\bf C} &    -0.52 &  0.13 & -0.58 &         -0.20 \\
% 		{\bf M} &    -0.59 & -0.59 & -0.61 &         -0.79 \\
% 		{\bf F} &    -0.11 & -0.08 & -0.13 &          0.41 \\
% 		{\bf I} &    -0.06 &  0.66 & -0.09 &          0.48 \\
% 		{\bf L} &     0.78 &  0.73 &  0.81 &          0.61 \\
% 		{\bf V} &     0.53 &  0.49 &  0.74 &          0.59 \\
% 		{\bf W} &     0.56 &  0.32 &  0.73 &          0.77 \\
% 		{\bf Y} &     0.49 &  0.58 &  0.58 &          0.56 \\
% 		{\bf A} &     0.08 & -0.25 & -0.02 &          0.14 \\
% 		{\bf G} &    -0.03 & -0.15 & -0.22 &         -0.33 \\
% 		{\bf T} &    -0.76 & -0.75 & -0.72 &         -0.91 \\
% 		{\bf S} &    -0.58 & -0.82 & -0.61 &         -0.96 \\
% 		{\bf N} &    -0.38 & -0.56 & -0.59 &         -0.76 \\
% 		{\bf Q} &    -0.70 & -0.75 & -0.76 &         -0.89 \\
% 		{\bf D} &    -0.79 & -0.58 & -0.83 &         -0.26 \\
% 		{\bf E} &     0.33 &  0.66 &  0.35 &          0.76 \\
% 		{\bf H} &    -0.48 & -0.27 & -0.48 &         -0.29 \\
% 		{\bf R} &     0.44 &  0.72 &  0.60 &          0.92 \\
% 		{\bf K} &     0.07 &  0.61 &  0.21 &          0.89 \\
% 		{\bf P} &     0.34 &  0.42 &  0.41 &          0.45 \\		% \bottomrule
% 	\end{tabular}
% 	\begin{flushleft} Temperature trends of the frequencies of individual amino acids in different datasets and subsets. 
% 	\end{flushleft}
% 	\label{tab:correlations}
% \end{table}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\end{document}

